agent-control/src/on_host/file_store.rs

```diff
@@ -0,0 +1,775 @@
+use std::{
+    io::{self, Error, ErrorKind},
+    path::{Path, PathBuf},
+    sync::RwLock,
+};
+
+use fs::{
+    LocalFile,
+    directory_manager::{DirectoryManager, DirectoryManagerFs},
+    file_reader::{FileReader, FileReaderError},
+    writer_file::FileWriter,
+};
+use serde::{Serialize, de::DeserializeOwned};
+use tracing::{debug, error, trace};
+
+use crate::{
+    agent_control::{
+        agent_id::AgentID,
+        defaults::{FOLDER_NAME_FLEET_DATA, FOLDER_NAME_LOCAL_DATA},
+    },
+    opamp::data_store::{OpAMPDataStore, StoreKey},
+};
+
+pub struct FileStore<F, D>
+where
+    D: DirectoryManager,
+    F: FileWriter + FileReader,
+{
+    directory_manager: D,
+    file_rw: F,
+    remote_dir: RwLock<RemoteDir>,
+    local_dir: LocalDir,
+}
+
+pub struct LocalDir(PathBuf);
+
+impl LocalDir {
+    pub fn get_local_file_path(&self, agent_id: &AgentID, key: &StoreKey) -> PathBuf {
+        self.0
+            .join(FOLDER_NAME_LOCAL_DATA)
+            .join(agent_id)
+            .join(build_config_name(key))
+    }
+}
+
+pub struct RemoteDir(PathBuf);
+
+impl RemoteDir {
+    pub fn get_remote_file_path(&self, agent_id: &AgentID, key: &StoreKey) -> PathBuf {
+        self.0
+            .join(FOLDER_NAME_FLEET_DATA)
+            .join(agent_id)
+            .join(build_config_name(key))
+    }
+}
+
+impl FileStore<LocalFile, DirectoryManagerFs> {
+    pub fn new_local_fs(local_dir: PathBuf, remote_dir: PathBuf) -> Self {
+        let remote_dir = RwLock::new(RemoteDir(remote_dir));
+        let local_dir = LocalDir(local_dir);
+        Self {
+            file_rw: LocalFile,
+            directory_manager: DirectoryManagerFs,
+            local_dir,
+            remote_dir,
+        }
+    }
+}
+
+impl<F, D> FileStore<F, D>
+where
+    D: DirectoryManager,
+    F: FileWriter + FileReader,
+{
+    pub fn new(file_rw: F, directory_manager: D, local_dir: PathBuf, remote_dir: PathBuf) -> Self {
+        let remote_dir = RwLock::new(RemoteDir(remote_dir));
+        let local_dir = LocalDir(local_dir);
+        Self {
+            file_rw,
+            directory_manager,
+            local_dir,
+            remote_dir,
+        }
+    }
+
+    // Load a file contents only if the file is present.
+    // If the file is not present there is no error nor file
+    fn load_file_if_present(&self, path: PathBuf) -> Result<Option<String>, FileReaderError> {
+        let values_result = self.file_rw.read(path.as_path());
+        match values_result {
+            Ok(res) => Ok(Some(res)),
+            Err(FileReaderError::FileNotFound(e)) => {
+                trace!("file not found! {e}");
+                // actively fallback to load local file
+                Ok(None)
+            }
+            Err(err) => {
+                error!("error loading file {}", path.display());
+                Err(err)
+            }
```

Comment:
Nit: Maybe it's easier to read if we handle the special case first?

```suggestion
        let result = self.file_rw.read(path);
        if let Err(FileReaderError::FileNotFound(ref e)) = result {
            trace!("file not found! {e}");
            return Ok(None);
        }

        result
            .map(Some)
            .inspect_err(|_| error!("error loading file {}", path.display()))
```

Reply:
I see it clearer with the `match` handling all cases, WDYT @newrelic/ac ?